apps/ex_wire/lib/ex_wire/struct/warp_queue.ex

```diff
@@ -0,0 +1,275 @@
+defmodule ExWire.Struct.WarpQueue do
+  @moduledoc """
+  `WarpQueue` maintains the current state of an active warp, this mean we will 
+  track the block_chunk hashes and state_chunk hashes given to us, so we can
+  request each from peers.
+
+  TODO: This will likely need to be updated to handle warping from more than
+        one direct peer.
+  """
+
+  alias Block.Header
+  alias Blockchain.{Block, Blocktree}
+  alias ExWire.Packet.SnapshotData.BlockChunk
+  alias ExWire.Packet.SnapshotManifest
+  alias MerklePatriciaTree.Trie
+
+  @type t :: %{
+          manifests: [SnapshotManifest.t()],
+          first_block_hash: EVM.hash() | nil,
+          current_block_number: integer() | nil,
+          current_block_hash: EVM.hash() | nil,
+          known_block_hashes: MapSet.t(EVM.hash()),
+          known_stash_hashes: MapSet.t(EVM.hash()),
+          block_chunk_requests: MapSet.t(EVM.hash()),
+          unprocessed_block_data: %{integer() => BlockData.t()}
+        }
+  defstruct [
+    :manifests,
+    :first_block_hash,
+    :current_block_number,
+    :current_block_hash,
+    :known_block_hashes,
+    :known_state_hashes,
+    :block_chunk_requests,
+    :unprocessed_block_data
+  ]
+
+  @doc """
+  Creates a new `WarpQueue` that 
+  """
+  def new() do
+    %__MODULE__{
+      manifests: [],
+      first_block_hash: nil,
+      current_block_number: nil,
+      current_block_hash: nil,
+      known_block_hashes: MapSet.new(),
+      known_state_hashes: MapSet.new(),
+      block_chunk_requests: MapSet.new(),
+      unprocessed_block_data: %{}
+    }
+  end
+
+  @spec new_manifest(t(), SnapshotManifest.manifest()) :: t()
+  def new_manifest(warp_queue, manifest) do
+    updated_known_block_hashes =
+      MapSet.union(
+        warp_queue.known_block_hashes,
+        MapSet.new(manifest.block_hashes)
+      )
+
+    updated_known_state_hashes =
+      MapSet.union(
+        warp_queue.known_state_hashes,
+        MapSet.new(manifest.state_hashes)
+      )
+
+    %{
+      warp_queue
+      | manifests: [manifest | warp_queue.manifests],
+        first_block_hash: List.first(manifest.block_hashes),
+        current_block_number: nil,
+        current_block_hash: nil,
+        known_block_hashes: updated_known_block_hashes,
+        known_state_hashes: updated_known_state_hashes
+    }
+  end
+
+  @doc """
+  When we receive a new block chunk, we want to remove it from requests
+  and add it to our processing queue.
+  """
+  @spec new_block_chunk(WarpQueue.t(), BlockChunk.t()) :: WarpQueue.t()
+  def new_block_chunk(warp_queue, block_chunk) do
+    next_warp_queue =
+      if !is_nil(warp_queue.first_block_hash) &&
```

Comment:
Is it worth extracting some functions to make this check clearer? Something like:

```ex
def new_block_chunk(warp_queue, block_chunk) do
  should_update_queue = first_block_present?(warp_queue) &&
    block_chunk_hash(block_chunk) == warp_queue.first_block_hash

  next_warp_queue = if should_update_queue do
    %{
      warp_queue |
        first_block_hash: nil,
        current_block_number: block_chunk.number,
        current_block_hash: block_chunk.hash
    }
  else
    warp_queue
  end
...

defp first_block_present?(warp_queue)
  !is_nil(warp_queue.first_block_hash)
end

defp block_chunk_hash(block_chunk)
  block_chunk
  |> BlockChunk.serialize()
  |> ExRLP.encode()
  |> ExthCrypto.Hash.Keccak.kec()
end
```

Reply:
Yup, I'm still.... working this one out. But once I do, for sure!